src/services/ExpressionBuilder.ts

```diff
@@ -0,0 +1,640 @@
+
+  /**
+   * Separator between filters with different keys or operators
+   */
+  filterSeparator?: string;
+
+  /**
+   * Prefix of the logQL expression
+   */
+  prefix?: string;
+
+  /**
+   * Sets if match filters join values together in a regex filter
+   * i.e. multiple filters with operators: '=' | '!=' => '=~' | '!~'
+   */
+  joinMatchFilters: boolean;
+
+  /**
+   * Sets if the values are JSON encoded
+   */
+  decodeFilters: boolean;
+}
+
+export class ExpressionBuilder {
+  private filters: AdHocFilterWithLabels[];
+  private options: Options;
+  private valueSeparator = 'or';
+
+  constructor(filters: AdHocFilterWithLabels[], options: Options = { joinMatchFilters: true, decodeFilters: false }) {
+    this.filters = filters;
+    this.options = options;
+    if (!this.options.debug) {
+      this.options.debug = getExpressionBuilderDebug();
+    }
+  }
+
+  /**
+   * Joins filters with same keys similar operators
+   * e.g. {level="info"}, {level="warn"} => {level=~"warn|info"}
+   */
+  public getJoinedLabelsFilters(): AdHocFilterWithLabels[] {
+    let { equalsFilters, notEqualsFilters, regexEqualFilters, regexNotEqualFilters } = this.getCombinedLabelFilters();
+    const adHocFilters: AdHocFilterWithLabels[] = [];
+    const filters = [equalsFilters, notEqualsFilters, regexEqualFilters, regexNotEqualFilters].filter(
+      (filter) => filter
+    );
+    filters.forEach((operatorFilters) => {
+      const joinedValues = this.joinCombinedFiltersValues(operatorFilters, '|');
+
+      for (const key in operatorFilters) {
+        const filter = operatorFilters[key];
+        const operator = filter.operator;
+        adHocFilters.push({ key, operator, value: joinedValues[key] });
+      }
+    });
+
+    return adHocFilters;
+  }
+
+  /**
+   * Returns logQL expression for AdHocFilterWithLabels[]
+   * Merges multiple include matches into regex
+   */
+  public getLabelsExpr(): string {
+    let {
+      equalsFilters,
+      notEqualsFilters,
+      regexEqualFilters,
+      regexNotEqualFilters,
+      ltFilters,
+      lteFilters,
+      gtFilters,
+      gteFilters,
+    } = this.getCombinedLabelFilters();
+
+    if (this.options.debug) {
+      console.info('combined filters after merge', {
+        equalsFilters,
+        notEqualsFilters,
+        regexEqualFilters,
+        regexNotEqualFilters,
+        ltFilters,
+        lteFilters,
+        gtFilters,
+        gteFilters,
+      });
+    }
+
+    const filtersString = this.buildLabelsLogQLFromFilters({
+      equalsFilters,
+      notEqualsFilters,
+      regexEqualFilters,
+      regexNotEqualFilters,
+      ltFilters,
+      lteFilters,
+      gtFilters,
+      gteFilters,
+    });
+
+    if (filtersString) {
+      // Append prefix if defined
+      return (this.options.prefix ?? '') + filtersString;
+    }
+
+    return '';
+  }
+
+  /**
+   * Returns merged filters separated by pipe
+   */
+  public getMetadataExpr(): string {
+    this.options = {
```

Comment:
Feel free to decide whether to chage it in this or a follow up PR, but I find this approach very unusual and misleading for anyone consuming this class: options are set by the constructor, but then when someone calls one of the methods, any option set when the instance was created is going to be silently overriden? Feels like it could use a refactor to change how these options are set and used. 

For example, instead of overwriting any previous option set by the constructor, make it an argument to `this.getLabelsExpr()`.

Reply:
Agreed, fixed in https://github.com/grafana/explore-logs/pull/1045/commits/60497085e08f09a1149a5e75300810dd1f8a0401

Reply:
Aaaaaand updated the type def in [9a91e17](https://github.com/grafana/explore-logs/pull/1045/commits/9a91e177fa785b5b818a075333cca455159d860f) :P

Reply:
Cool, thanks! I just checked and it seems that ExpressionBuilder is not being instantiated with custom options, so maybe remove them from the constructor? I still feel that exposing them through the constructor and then overwriting them with calls to different methods defeats the purpose of them being instance options. Maybe it should be something internal that changes depending on what expression needs to be generated.
Not a huge deal, on the nitpicky side, but I think it would make a more consistent implementation, specially if stuff like the filterSeparator doesn't really change.

Reply:
Yeah, but then we have to make the options object optional everywhere if it's not set in the constructor.

I would argue the class is built for more generic use-cases that we don't currently have implemented, and the public methods besides `getLabelsExpr` and `getJoinedLabelsFilters` are entirely unnecessary syntactic sugar methods that just keep the options out of the generic constructor.

TL;DR if you want to set options in the constructor, call `getLabelsExpr` or `getJoinedLabelsFilters` directly.

If we really don't like overwriting the options in the public helper methods then we should export default options for specific use-cases (METADATA_PARSING_OPTIONS, FIELD_PARSING_OPTIONS), and always call the generic methods.

Reply:
Maybe most of what I see is a byproduct of being an instantiable class. I was just thinking that by the way it works, every public method could be static and have the filters and options be arguments.

Reply:
Agreed that would be a simpler API, but I'm not sure the juice is worth the squeeze with the current complexity of the class. But something to look at when we migrate the pattern filters or whatever else gets added next.